core/src/main/scala/chisel3/internal/Builder.scala

```diff
@@ -1160,6 +1162,36 @@ private[chisel3] object Builder extends LazyLogging {
       )
     }
   }
+
+  // Puts a module prefix string onto the module prefix stack
+  def pushModulePrefix(prefix: String): Unit = {
+    val context = chiselContext.get()
+    context.modulePrefixStack = prefix :: context.modulePrefixStack
+  }
+
+  // Remove the module prefix on top of the stack
+  def popModulePrefix(): List[String] = {
+    val context = chiselContext.get()
+    val tail = context.modulePrefixStack.tail
+    context.modulePrefixStack = tail
+    tail
+  }
+
+  // Returns the nested module prefix at this moment
+  def getModulePrefix: String = {
+    val modulePrefixStack = chiselContext.get().modulePrefixStack
+    if (modulePrefixStack.isEmpty) {
+      ""
+    } else {
+      modulePrefixStack.foldLeft("")((a, b) => b + "_" + a)
```

Comment:
I'm of two minds on automatically including `_` between prefixes. On the one hand, that is almost certainly the most common use, but on the other hand, what if a user doesn't want `_` between prefixes?

I'm thinking that instead of adding them here, maybe we overload `withModulePrefix`, something like:
```scala
object withModulePrefix {
  def apply[T](prefix: String, trailingUnderscore: Boolean)(block: => T): T = ??? 
  
  def apply[T](prefix: String)(block: => T): T = apply(prefix, true)(block)
```
So that the default behavior adds the trailing `_` but the user can opt-out? WDYT?

Reply:
It strikes me as awkward to mandate the `_` as part of the prefix. You're going to get someone who forgets it, and it's gonna make things confusing and ugly.

I wouldn't be against a global-level configuration for the separator, with `"_"` the default value for it.

Reply:
I don't think we need this. Users can add _ if they want, right?

Reply:
I did an audit of our internal code and AFAICT every single place except one[^1] the added prefix ends with `_`. I think it's reasonable to be opinionated that separators are a good thing and to default to `_`. If it comes up that some users either want a different separator, or none at all--then we can make it possible to customize the separator when invoking Chisel, and potentially make it possible to customize at specific call sites.

@mmaloney-sf one thing that having separators by default does imply though is that we need to specially handle empty Strings as not adding any prefix (i.e. don't add a separator in that case).

[^1]: And that one place it might be a mistake to not have the separator.